src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -437,29 +523,109 @@ HeapWord* ShenandoahFreeSet::try_allocate_in(ShenandoahHeapRegion* r, Shenandoah
+        ShenandoahHeap::fill_with_object(fill_addr, fill_size);
+        r->set_top(r->end());
+        // Since we have filled the waste with an empty object, account for increased usage
+        _heap->increase_used(waste);
+      } else {
+        // We'll retire the region until the freeset is rebuilt. Since retiement is not permanent, we do not account for waste.
+        waste = 0;
+      }
+      if (probe_old_collector_set(idx)) {
+        assert(_heap->mode()->is_generational(), "Old collector free regions only present in generational mode");
+        if (waste > 0) {
+          _heap->old_generation()->increase_used(waste);
+          _heap->card_scan()->register_object(fill_addr);
+        }
+        remove_from_old_collector_set(idx);
+        assert(!in_collector_set(idx) && !in_mutator_set(idx), "Region cannot be in multiple free sets");
+        adjust_old_collector_bounds_if_touched(idx);
+      } else {
+        assert(probe_collector_set(idx), "Region that is not mutator free must be collector free or old collector free");
+        if ((waste > 0) && _heap->mode()->is_generational()) {
+          _heap->young_generation()->increase_used(waste);
+        }
+        // This applies to both generational and non-generational mode
+        remove_from_collector_set(idx);
+        assert(!in_mutator_set(idx) && !in_old_collector_set(idx), "Region cannot be in multiple free sets");
+        adjust_collector_bounds_if_touched(idx);
+      }
     }
     assert_bounds();
   }
   return result;
 }
 
+// If idx represents a mutator bound, recompute the mutator bounds, returning true iff bounds were adjusted.
+bool ShenandoahFreeSet::adjust_mutator_bounds_if_touched(size_t idx) {
+  if (idx == _mutator_leftmost || idx == _mutator_rightmost) {
+    // Rewind both mutator bounds until the next bit.
+    while (_mutator_leftmost < _max && !in_mutator_set(_mutator_leftmost)) {
+      _mutator_leftmost++;
+    }
+    while (_mutator_rightmost > 0 && !in_mutator_set(_mutator_rightmost)) {
+      _mutator_rightmost--;
+    }
+    return true;
+  } else {
+    return false;
+  }
+}
+
+// If idx represents an old collector bound, recompute the old collector bounds, returning true iff bounds were adjusted.
+bool ShenandoahFreeSet::adjust_old_collector_bounds_if_touched(size_t idx) {
+  if (idx == _old_collector_leftmost || idx == _old_collector_rightmost) {
+    // Rewind both old collector bounds until the next bit.
+    while (_old_collector_leftmost < _max && !in_old_collector_set(_old_collector_leftmost)) {
+      _old_collector_leftmost++;
+    }
+    while (_old_collector_rightmost > 0 && !in_old_collector_set(_old_collector_rightmost)) {
+      _old_collector_rightmost--;
+    }
+    return true;
+  } else {
+    return false;
+  }
+}
+
+// If idx represents a collector bound, recompute the collector bounds, returning true iff bounds were adjusted.
+bool ShenandoahFreeSet::adjust_collector_bounds_if_touched(size_t idx) {
```

Comment:
This really feels like there should be a `RegionFreeSet` class with instances for `mutator`, `collector`, and `old_collector`. There's a lot of code duplication here. A refactoring for another PR.

Reply:
Agree that refactoring would be nice in the future.

Reply:
I've gone partway there, but more thorough refactoring can happen in a future PR.